app/services/search_service.rb

```diff
@@ -35,12 +35,61 @@ def filter_conversations
   end
 
   def filter_messages
-    @messages = current_account.messages.where(inbox_id: accessable_inbox_ids)
-                               .where('messages.content ILIKE :search', search: "%#{search_query}%")
-                               .where('created_at >= ?', 3.months.ago)
-                               .reorder('created_at DESC')
-                               .page(params[:page])
-                               .per(15)
+    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+    @messages = if use_gin_search
+                  filter_messages_with_gin
+                else
+                  filter_messages_with_like
+                end
+
+    log_search_performance(start_time)
+    @messages
+  end
+
+  def filter_messages_with_gin
+    base_query = message_base_query
+
+    if search_query.present?
+      # Use the @@ operator with to_tsquery for better GIN index utilization
+      # Convert search query to tsquery format with prefix matching
+      tsquery = search_query.split.map { |word| "#{word}:*" }.join(' & ')
```

Comment:
Note: The search results will be different in GIN compared to ILIKE since we are splitting words and combining using `&` filtering in GIN. GIN would get us exact matches but ILIKE would have more results.